File: EclipseEmu/Emulation/RingBuffer.swift
import Atomics
import Foundation

struct RingBuffer: ~Copyable {
    let capacity: Int
    private var head: ManagedAtomic<Int> = .init(0)
    private var tail: ManagedAtomic<Int> = .init(0)
    private var inner: UnsafeMutableRawBufferPointer

    init(capacity: Int) {
        self.capacity = capacity
        self.inner = UnsafeMutableRawBufferPointer.allocate(
            byteCount: MemoryLayout<UInt8>.stride * capacity,
            alignment: MemoryLayout<UInt8>.alignment
        )
    }

    deinit {
        self.inner.deallocate()
    }

    @inlinable
    func availableRead(head: Int, tail: Int) -> Int {
        return tail >= head ? tail &- head : tail &+ self.capacity &- head
    }

    @inlinable
    func availableWrite(head: Int, tail: Int) -> Int {
        return tail >= head ? self.capacity &- tail &+ head : head &- tail
    }

    func availableRead() -> Int {
        let head = self.head.load(ordering: .relaxed)
        let tail = self.tail.load(ordering: .relaxed)
        return self.availableRead(head: head, tail: tail)
    }

    func availableWrite() -> Int {
        let head = self.head.load(ordering: .relaxed)
        let tail = self.tail.load(ordering: .relaxed)
        return self.availableWrite(head: head, tail: tail)
    }

    mutating func read(dst: UnsafeMutableRawPointer, length: Int) -> Int {
        let head = self.head.load(ordering: .acquiring)
        let tail = self.tail.load(ordering: .relaxed)

        let available = self.availableRead(head: head, tail: tail)
        guard available >= length, let src = self.inner.baseAddress else { return 0 }

        var nextHead = head + length
        let needsWrap = nextHead >= self.capacity
        nextHead -= needsWrap ? self.capacity : 0
        self.head.store(nextHead, ordering: .releasing)

        let len1 = needsWrap ? self.capacity - head : length
        let len2 = needsWrap ? nextHead : 0

        memcpy(dst, src.advanced(by: head), len1)
        memcpy(dst.advanced(by: len1), src, len2)

        return length
    }

    mutating func write(src: UnsafeRawPointer, length: Int) -> Int {
        let head = self.head.load(ordering: .relaxed)
        let tail = self.tail.load(ordering: .acquiring)

        let available = self.availableWrite(head: head, tail: tail)
        guard available >= length, let dst = self.inner.baseAddress else { return 0 }

        var nextTail = tail + length
        let needsWrap = nextTail >= self.capacity
        nextTail -= needsWrap ? self.capacity : 0
        self.tail.store(nextTail, ordering: .releasing)

        let len1 = needsWrap ? self.capacity &- tail : length
        let len2 = needsWrap ? nextTail : 0

        memcpy(dst.advanced(by: tail), src, len1)
        memcpy(dst, src.advanced(by: len1), len2)

        return length
    }

    func clear() {
        self.head.store(0, ordering: .relaxed)
        self.tail.store(0, ordering: .relaxed)
    }
}

File: EclipseEmu/Emulation/RingBuffer.c
#include "RingBuffer.h"
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

RingBuffer *ring_buffer_init(uint64_t capacity) {
    struct RingBuffer *ring = malloc(sizeof(struct RingBuffer));
    if (ring == NULL) {
        return NULL;
    }
    
    void *inner = malloc(capacity);
    if (inner == NULL) {
        free(ring);
        return NULL;
    }
    
    ring->inner = inner;
    ring->capacity = capacity;
    ring->head = 0;
    ring->tail = 0;

    return ring;
}

inline void ring_buffer_deinit(RingBuffer *self) {
    free(self->inner);
    free(self);
}

inline uint64_t ring_buffer_available_read_preloaded(RingBuffer *self, uint64_t tail, uint64_t head) {
    // NOTE: this is equivalent to
    // `return tail >= head ? tail - head : tail + self->capacity - head;`
    return tail + ((tail >= head) * self->capacity) - head;
}

inline uint64_t ring_buffer_available_write_preloaded(RingBuffer *self, uint64_t tail, uint64_t head) {
    return tail >= head ? self->capacity - tail + head : head - tail;
}

inline uint64_t ring_buffer_available_read(RingBuffer *self) {
    uint64_t head = atomic_load_explicit(&(self->head), memory_order_relaxed);
    uint64_t tail = atomic_load_explicit(&(self->tail), memory_order_relaxed);
    return ring_buffer_available_read_preloaded(self, tail, head);
}

inline uint64_t ring_buffer_available_write(RingBuffer *self) {
    uint64_t head = atomic_load_explicit(&(self->head), memory_order_relaxed);
    uint64_t tail = atomic_load_explicit(&(self->tail), memory_order_relaxed);
    return ring_buffer_available_write_preloaded(self, tail, head);
}

/// this is under the assumption that length will never be more than capacity
uint64_t ring_buffer_write(RingBuffer *self, const void *src, uint64_t length) {
    uint64_t head = atomic_load_explicit(&(self->head), memory_order_relaxed);
    uint64_t tail = atomic_load_explicit(&(self->tail), memory_order_acquire);
    
    uint64_t available = ring_buffer_available_write_preloaded(self, tail, head);
    if (available < length) {
        return 0;
    }
    
    uint64_t next_tail = tail + length;
    bool needs_wrap = next_tail >= self->capacity;
    next_tail -= needs_wrap * self->capacity;
    
    uint64_t len1 = needs_wrap ? self->capacity - tail : length;
    uint64_t len2 = needs_wrap * next_tail;
    
    memcpy(self->inner + tail, src, len1);
    memcpy(self->inner, src + len1, len2);
    
    atomic_store_explicit(&(self->tail), next_tail, memory_order_release);
    return length;
}

uint64_t ring_buffer_read(RingBuffer *self, void *dst, uint64_t length) {
    uint64_t head = atomic_load_explicit(&(self->head), memory_order_acquire);
    uint64_t tail = atomic_load_explicit(&(self->tail), memory_order_relaxed);
    
    uint64_t available = ring_buffer_available_read_preloaded(self, tail, head);
    if (available < length) {
        return 0;
    }
    
    uint64_t next_head = head + length;
    bool needs_wrap = next_head >= self->capacity;
    next_head -= needs_wrap * self->capacity;
    
    uint64_t len1 = needs_wrap ? self->capacity - head : length;
    uint64_t len2 = needs_wrap * next_head;
    
    memcpy(dst, self->inner + head, len1);
    memcpy(dst + head, self->inner, len2);
    
    atomic_store_explicit(&(self->head), next_head, memory_order_release);
    return length;
}

void ring_buffer_clear(RingBuffer *self) {
    atomic_store_explicit((&self->head), 0, memory_order_relaxed);
    atomic_store_explicit((&self->tail), 0, memory_order_relaxed);
    memset(self->inner, 0, self->capacity);
}
